presto-main/src/main/java/com/facebook/presto/operator/scalar/TDigest.java

```diff
@@ -0,0 +1,182 @@
+ * limitations under the License.
+ */
+
+package com.facebook.presto.operator.scalar;
+
+import com.google.common.base.Preconditions;
+import io.airlift.slice.Slice;
+
+import java.io.Serializable;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Random;
+
+public abstract class TDigest
+        implements Serializable
+{
+    protected ScaleFunction scale = ScaleFunction.K_2;
+    double min = Double.POSITIVE_INFINITY;
+    double max = Double.NEGATIVE_INFINITY;
+
+    final Random random = new Random();
+    boolean recordAllData;
+
+    @SuppressWarnings("WeakerAccess")
+    public static TDigest createMergingDigest(double compression)
+    {
+        return new MergingDigest(compression);
+    }
+
+    public abstract void add(double value, int weight);
+
+    abstract void add(double value, int weight, Centroid base);
+
+    public void add(double value)
+    {
+        add(value, 1);
+    }
+
+    public void add(TDigest other)
+    {
+        List<Centroid> tmp = new ArrayList<>();
+        for (Centroid centroid : other.centroids()) {
+            tmp.add(centroid);
+        }
+
+        Collections.shuffle(tmp, random);
+        for (Centroid centroid : tmp) {
+            add(centroid.mean(), centroid.count(), centroid);
+        }
+    }
+
+    final void checkValue(double x)
+    {
+        if (Double.isNaN(x)) {
+            throw new IllegalArgumentException("Cannot add NaN");
+        }
+    }
+
+    /**
+     * Re-examines a t-digest to determine whether some centroids are redundant.  If your data are
+     * perversely ordered, this may be a good idea.  Even if not, this may save 20% or so in space.
+     *
+     * This is a destructive operation that is not thread-safe.
+     */
+    public abstract void compress();
+
+    public abstract long size();
+
+    /**
+     * Returns the fraction of all points added which are &le; x.
+     *
+     * @param x The cutoff for the cdf.
+     * @return The fraction of all data which is less or equal to x.
+     */
+    public abstract double cdf(double x);
+
+    /**
+     * Returns an estimate of the cutoff such that a specified fraction of the data
+     * added to this TDigest would be less than or equal to the cutoff.
+     *
+     * @param q The desired fraction
+     * @return The value x such that cdf(x) == q
+     */
+    public abstract double quantile(double q);
+
+    /**
+     * A {@link Collection} that lets you go through the centroids in ascending order by mean.  Centroids
+     * returned will not be re-used, but may or may not share storage with this TDigest.
+     *
+     * @return The centroids in the form of a Collection.
+     */
+    public abstract Collection<Centroid> centroids();
+
+    /**
+     * Returns the current compression factor.
+     *
+     * @return The compression factor originally used to set up the TDigest.
```

Comment:
We try to avoid these sorts of comments where one part describes the method and the next part uses javadoc notation to describe the exact same thing